### ISSUE TITLE
Deprecate bytes and metadata as numpy arrays in datasets

### ISSUE THREAD

**Issue by NicolasGensollen** [nilearn/nilearn#2660]
Most fetchers in `nilearn.datasets` currently return descriptions and label names in `bytes`. See issue #1635 and recent PR #2655 for example.
Moreover, metadata is currently returned as numpy arrays which is not optimal from a usage perspective. 
For example:

```python
from nilearn.datasets import fetch_adhd
dset = fetch_adhd()
print(type(dset['phenotypic']))
print(dset['phenotypic'][0])
```
 outputs

```
<class 'numpy.ndarray'>
(b'"21"', 10042, b'"rest_1"', 0.0559, 0, 0.2365, 0.0922, 0., 2.2915, 1.0089, b'"NYU"', b'NA', b'"data_set"', 10.65, b'"M"', b'"0.91"', b'NA', b'108', b'100', b'115', b'2', 0, 1, b'NA', b'NA', b'NA', b'""', b'""', b'""', b'NA', b'NA', b'NA', b'NA', b'NA', b'NA', b'NA', b'NA', b'NA', b'NA', b'NA', b'59', b'56', b'65', b'NA', b'"pass"', b'""', b'"pass"', b'""', b'""', b'""', b'NA', b'NA', b'NA', b'NA', b'NA', b'NA', b'"pass"', b'NA', b'NA', b'NA', b'NA', b'NA', b'NA', b'""', b'""')
```

Since we were planning to use Pandas instead of Numpy for this (see #2621), I think this could be the opportunity to update all fetchers to return proper python strings and pandas dataframes. 

Since this would be a breaking change, we should add a deprecation message in the datasets module, and aim for releasing these changes in version `0.9`.

**Comment by bthirion** [comment]
+1 for introducing the deprecation warning.
So you will do the conversions (numpy-> pandas, bytes->string in the fetcher) ?

**Comment by NicolasGensollen** [comment]
@bthirion OK, I'll add a deprecation warning asap.

>So you will do the conversions (numpy-> pandas, bytes->string in the fetcher) ?

Yes, I'm planning to do that.

**Comment by Remi-Gau** [comment]
script to check types of the content of the dataset

will try to use something similar in the test suite, but not sure we'll catch them all given we are mocking some data

```python
from nilearn.surface.surface import SurfaceImage, PolyMesh
from nibabel import Nifti1Image
from sklearn.utils import Bunch
import pandas as pd
from rich import print

from nilearn.datasets.struct import (
    fetch_icbm152_2009,
    fetch_icbm152_brain_gm_mask,
    fetch_oasis_vbm,
    fetch_surf_fsaverage,
    load_fsaverage,
    load_fsaverage_data,
    load_mni152_brain_mask,
    load_mni152_gm_mask,
    load_mni152_gm_template,
    load_mni152_template,
    load_mni152_wm_mask,
    load_mni152_wm_template,
)

from nilearn.datasets.func import (
    fetch_abide_pcp,
    fetch_adhd,
    fetch_development_fmri,
    fetch_ds000030_urls,
    fetch_fiac_first_level,
    fetch_haxby,
    fetch_language_localizer_demo_dataset,
    fetch_localizer_button_task,
    fetch_localizer_calculation_task,
    fetch_localizer_contrasts,
    fetch_localizer_first_level,
    fetch_megatrawls_netmats,
    fetch_mixed_gambles,
    fetch_miyawaki2008,
    fetch_spm_auditory,
    fetch_spm_multimodal_fmri,
    fetch_surf_nki_enhanced,
    load_nki,
    load_sample_motor_activation_image,
)


def check_type(data):
    if isinstance(
        data,
        (str, int, float, Nifti1Image, SurfaceImage, pd.DataFrame, PolyMesh),
    ):
        pass
    elif isinstance(data, (Bunch, dict)):
        for k, v in data.items():
            if not check_type(v):
                print(f"{k} : {type(v)}")
    elif isinstance(data, (set, list, tuple)):
        for v in data:
            if not check_type(v):
                print(f"{type(v)}")
    else:
        return False
    return True


for f in [
    # struct
    fetch_icbm152_2009,
    fetch_icbm152_brain_gm_mask,
    fetch_surf_fsaverage,
    load_fsaverage,
    load_fsaverage_data,
    load_mni152_brain_mask,
    load_mni152_gm_mask,
    load_mni152_gm_template,
    load_mni152_template,
    load_mni152_wm_mask,
    load_mni152_wm_template,
    # func
    fetch_fiac_first_level,
    fetch_haxby,
    fetch_localizer_button_task,
    fetch_localizer_first_level,
    fetch_megatrawls_netmats,
    fetch_miyawaki2008,
    fetch_spm_auditory,
    fetch_spm_multimodal_fmri,
    load_nki,
    load_sample_motor_activation_image,
    fetch_ds000030_urls,
]:
    data = f()
    check_type(data)

for f in [
    fetch_adhd,
    fetch_surf_nki_enhanced,
    fetch_mixed_gambles,
    fetch_localizer_calculation_task,
    fetch_development_fmri,
    fetch_oasis_vbm,
]:
    data = f(n_subjects=1)
    check_type(data)

data = fetch_language_localizer_demo_dataset(legacy_output=False)
check_type(data)

fetch_localizer_contrasts(
    ["checkerboard"],
    n_subjects=1,
    get_tmaps=True,
    get_masks=True,
    get_anats=True,
)
check_type(data)

data = fetch_abide_pcp(
    n_subjects=1,
    band_pass_filtering=True,
    global_signal_regression=True,
    derivatives=["func_preproc"],
    quality_checked=True,
)
check_type(data)

```
